src/provider/provider_devdax_memory.c

```diff
@@ -0,0 +1,677 @@
+
+static __TLS os_last_native_error_t TLS_last_native_error;
+
+// helper values used only in the Native_error_str array
+#define _UMF_DEVDAX_RESULT_SUCCESS                                             \
+    (UMF_DEVDAX_RESULT_SUCCESS - UMF_DEVDAX_RESULT_SUCCESS)
+#define _UMF_DEVDAX_RESULT_ERROR_ALLOC_FAILED                                  \
+    (UMF_DEVDAX_RESULT_ERROR_ALLOC_FAILED - UMF_DEVDAX_RESULT_SUCCESS)
+#define _UMF_DEVDAX_RESULT_ERROR_ADDRESS_NOT_ALIGNED                           \
+    (UMF_DEVDAX_RESULT_ERROR_ADDRESS_NOT_ALIGNED - UMF_DEVDAX_RESULT_SUCCESS)
+#define _UMF_DEVDAX_RESULT_ERROR_FREE_FAILED                                   \
+    (UMF_DEVDAX_RESULT_ERROR_FREE_FAILED - UMF_DEVDAX_RESULT_SUCCESS)
+#define _UMF_DEVDAX_RESULT_ERROR_PURGE_LAZY_FAILED                             \
+    (UMF_DEVDAX_RESULT_ERROR_PURGE_LAZY_FAILED - UMF_DEVDAX_RESULT_SUCCESS)
+#define _UMF_DEVDAX_RESULT_ERROR_PURGE_FORCE_FAILED                            \
+    (UMF_DEVDAX_RESULT_ERROR_PURGE_FORCE_FAILED - UMF_DEVDAX_RESULT_SUCCESS)
+
+static const char *Native_error_str[] = {
+    [_UMF_DEVDAX_RESULT_SUCCESS] = "success",
+    [_UMF_DEVDAX_RESULT_ERROR_ALLOC_FAILED] = "memory allocation failed",
+    [_UMF_DEVDAX_RESULT_ERROR_ADDRESS_NOT_ALIGNED] =
+        "allocated address is not aligned",
+    [_UMF_DEVDAX_RESULT_ERROR_FREE_FAILED] = "memory deallocation failed",
+    [_UMF_DEVDAX_RESULT_ERROR_PURGE_LAZY_FAILED] = "lazy purging failed",
+    [_UMF_DEVDAX_RESULT_ERROR_PURGE_FORCE_FAILED] = "force purging failed",
+};
+
+static void os_store_last_native_error(int32_t native_error, int errno_value) {
+    TLS_last_native_error.native_error = native_error;
+    TLS_last_native_error.errno_value = errno_value;
+}
+
+static int devdax_copy_path(const char *in_path, char out_path[NAME_MAX]) {
+    // (- 1) because there should be a room for the terminating null byte ('\0')
+    size_t max_len = NAME_MAX - 1;
+
+    if (strlen(in_path) > max_len) {
+        LOG_ERR("path of a devdax is longer than %zu bytes", max_len);
+        return -1;
+    }
+
+    strncpy(out_path, in_path, max_len);
+    out_path[NAME_MAX - 1] = '\0'; // the terminating null byte
+
+    return 0;
+}
+
+static umf_result_t
+devdax_translate_params(umf_devdax_memory_provider_params_t *in_params,
+                        devdax_memory_provider_t *provider) {
+    umf_result_t result;
+
+    result = os_translate_mem_protection_flags(in_params->protection,
+                                               &provider->protection);
+    if (result != UMF_RESULT_SUCCESS) {
+        LOG_ERR("incorrect memory protection flags: %u", in_params->protection);
+        return result;
+    }
+
+    /* devdax requires UMF_MEM_MAP_SHARED */
+    result = os_translate_mem_visibility_flag(UMF_MEM_MAP_SHARED,
+                                              &provider->visibility);
+    if (result != UMF_RESULT_SUCCESS) {
+        LOG_ERR("incorrect memory visibility flag: %u", UMF_MEM_MAP_SHARED);
```

Comment:
Translation of `UMF_MEM_MAP_SHARED` flag can probably only fail due to internal bug. `UMF_MEM_MAP_SHARED` doesn't come from the user-provided params, so an assertion might be more appropriate.

Reply:
Done

Reply:
@kswiecicki Please verify and resolve.